[C++ API] Allow skipping default arguments in module's forward method when module is used in Sequential

yf225 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33027 [C++ API] Allow module to populate default arguments in its forward method when used in a Sequential module**
* #33026 [C++ API] Expose AnyValue and AnyModuleHolder classes

This PR allows default arguments in module's forward method to be skipped when module is used in `torch::nn::Sequential`, by introducing the `FORWARD_HAS_DEFAULT_ARGS` macro and requiring that all modules that have default arguments in its forward method must have a corresponding `FORWARD_HAS_DEFAULT_ARGS` macro call.

Fixes issue mentioned in https://github.com/pytorch/pytorch/issues/30931#issuecomment-564144468.

Differential Revision: [D19777815](https://our.internmc.facebook.com/intern/diff/D19777815)

Commit:
[C++ API] Allow module to populate default arguments in its forward method when used in a Sequential module

Commit:
Update on "[C++ API] Allow module to populate default arguments in its forward method when used in a Sequential module"

[C++ API] Allow module to populate default arguments in its forward method when used in a Sequential module

gh-metadata: pytorch pytorch 33027 gh/yf225/95/head

Commit:
Update on "[C++ API] Allow module to populate default arguments in its forward method when used in a Sequential module"

[C++ API] Allow module to populate default arguments in its forward method when used in a Sequential module

gh-metadata: pytorch pytorch 33027 gh/yf225/95/head

Commit:
Update on "[C++ API] Allow skipping default arguments in module's forward method when module is used in Sequential"

[C++ API] Allow module to populate default arguments in its forward method when used in a Sequential module

gh-metadata: pytorch pytorch 33027 gh/yf225/95/head

pbelevich reviewed (commented):
```
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] AnyModuleTest.WrongNumberOfArguments
[  FAILED  ] AnyModuleTest.PassingArgumentsToModuleWithDefaultArgumentsInForwardMethod
```

Commit:
Update on "[C++ API] Allow skipping default arguments in module's forward method when module is used in Sequential"

[C++ API] Allow module to populate default arguments in its forward method when used in a Sequential module

gh-metadata: pytorch pytorch 33027 gh/yf225/95/head

Commit:
Update on "[C++ API] Allow skipping default arguments in module's forward method when module is used in Sequential"

[C++ API] Allow module to populate default arguments in its forward method when used in a Sequential module

gh-metadata: pytorch pytorch 33027 gh/yf225/95/head

Commit:
Update on "[C++ API] Allow skipping default arguments in module's forward method when module is used in Sequential"

[C++ API] Allow module to populate default arguments in its forward method when used in a Sequential module

gh-metadata: pytorch pytorch 33027 gh/yf225/95/head

Commit:
Update on "[C++ API] Allow skipping default arguments in module's forward method when module is used in Sequential"

[C++ API] Allow module to populate default arguments in its forward method when used in a Sequential module

gh-metadata: pytorch pytorch 33027 gh/yf225/95/head

Commit:
Update on "[C++ API] Allow skipping default arguments in module's forward method when module is used in Sequential"

[C++ API] Allow module to populate default arguments in its forward method when used in a Sequential module

gh-metadata: pytorch pytorch 33027 gh/yf225/95/head

Commit:
Update on "[C++ API] Allow skipping default arguments in module's forward method when module is used in Sequential"

[C++ API] Allow module to populate default arguments in its forward method when used in a Sequential module

gh-metadata: pytorch pytorch 33027 gh/yf225/95/head

Commit:
Update on "[C++ API] Allow skipping default arguments in module's forward method when module is used in Sequential"

[C++ API] Allow module to populate default arguments in its forward method when used in a Sequential module

gh-metadata: pytorch pytorch 33027 gh/yf225/95/head

Commit:
Update on "[C++ API] Allow skipping default arguments in module's forward method when module is used in Sequential"

[C++ API] Allow module to populate default arguments in its forward method when used in a Sequential module

gh-metadata: pytorch pytorch 33027 gh/yf225/95/head

Commit:
Update on "[C++ API] Allow skipping default arguments in module's forward method when module is used in Sequential"

[C++ API] Allow module to populate default arguments in its forward method when used in a Sequential module

gh-metadata: pytorch pytorch 33027 gh/yf225/95/head

pbelevich reviewed (approved):
nice work!

Commit:
Update on "[C++ API] Allow skipping default arguments in module's forward method when module is used in Sequential"

[C++ API] Allow module to populate default arguments in its forward method when used in a Sequential module

gh-metadata: pytorch pytorch 33027 gh/yf225/95/head

Commit:
Update on "[C++ API] Allow skipping default arguments in module's forward method when module is used in Sequential"

[C++ API] Allow module to populate default arguments in its forward method when used in a Sequential module

gh-metadata: pytorch pytorch 33027 gh/yf225/95/head
